sdks/python/hatchet_sdk/worker/runner/runner.py

```diff
@@ -67,6 +68,114 @@ class WorkerStatus(Enum):
     UNHEALTHY = 4
 
 
+@dataclass
+class _BatchQueueItem:
+    ctx: Context
+    future: "asyncio.Future[Any]"
+
+
+@dataclass
+class _BatchState:
+    expected_size: int
+    items: dict[int, _BatchQueueItem]
+    started: bool
+    trigger_reason: str | None = None
+    trigger_time: datetime | None = None
+    batch_key: str | None = None
+
+
+class _BatchController:
+    def __init__(self, action_id: str, task: Task[Any, Any]) -> None:
+        self.action_id = action_id
+        self.task = task
+        self.batches: dict[str, _BatchState] = {}
+
+    def enqueue_item(
+        self,
+        *,
+        batch_id: str,
+        expected_size: int,
+        index: int,
+        ctx: Context,
+        future: "asyncio.Future[Any]",
+        batch_key: str | None = None,
+        on_ready: Callable[[str], None],
+    ) -> None:
+        state = self.batches.get(batch_id)
+
+        if state is None:
+            state = _BatchState(
+                expected_size=expected_size,
+                items={},
+                started=False,
+                batch_key=batch_key,
+            )
+            self.batches[batch_id] = state
+        elif state.expected_size != expected_size and expected_size > 0:
+            raise ValueError(
+                f"Batch metadata mismatch for '{self.action_id}' batch {batch_id}: expected size {state.expected_size}, got {expected_size}"
+            )
```

Comment:
what happens if this raises? do the tasks ever start? or does the engine find out somehow?